### ISSUE TITLE
Spike: investigate feasibility of event-driven weighted reactions

### ISSUE THREAD

**Issue by jnaviask** [hicommonwealth/commonwealth#5553]
## Description

The Community Staking project involves the implementation of a new feature: weighted reactions. Rather than a reaction counting for "1 point", reaction counting when staking is enabled will vary depending on the value of a some data element relating to the reacting user. Importantly, this value is subject to change over time, as the user buys or sells stake in the community.

Assumptions:
- When a user's stake changes, an "Update Stake" command is run against the user model containing the address and stake bought/sold. The design of this piece is dealt with in #5552 and is out of scope for this ticket.
- The "Update Stake" command should trigger an update to the user's group membership in the community. We can assume that a group already exists representing the staking token. The membership should ideally reflect the user's total stake. We may in the future have multiple "strategies" for updating membership points, which will not be exact reflections of the onchain point total, so we should ensure that this design of this update and storage is sufficiently abstract to support multiple possible "pointing policies".
- Updating Stake should trigger an event which triggers an Update Reactions policy, that updates the weight of the user's reactions in accordance with the change in staked token value. We should take into careful consideration the database load involved in bulk updating reaction values, to ensure that what we build will scale for users with reaction counts in the 1000s.
- We should also take into consideration any API changes required in terms of querying reactions, to support weight values if they exist.

The scope of this spike is to write up a brief proposal of changes required to implement the Update Stake command and Update Reactions policy (i.e. the procedure which accepts the change in membership points and triggers updates on reactions) according to the above set of assumptions. This will ultimately become the basis for an implementation ticket as we move forward with Community Staking.

A diagram of the event flow involved in this spike:
![2023-11-02-104447_1828x405_scrot](https://github.com/hicommonwealth/commonwealth/assets/1860629/bc4b473d-7172-425a-9f12-a31e8803d51c)

## Project Owner

@jnaviask 

## Additional context

Timebox: 3 hours.

**Comment by rbennettcw** [comment]
@jnaviask I'm trying to understand how stakes will be modeled in the DB.

Since you said "the membership should ideally reflect the user's total stake", it sounds like the plan is to use something like `Memberships.stake_amount`.

The issue I'm seeing is that if a user is reacting to a thread with Topic `T`, and if Topic `T` is gated by Group `A` and Group `B`, and the user address is accepted in both groups, then there are multiple memberships (`MA`, `MB`) and thus multiple stake amounts associated with the address and that thread or comment being reacted to.

Does this make sense, and is it an issue?

Would be great if I could see a UI rendering of how weighted reactions is supposed to look to the end-user.

---

I've drafted a potential solution to support multiple pointing strategies, but first need to clarify how we'll handle the multiple-groups-per-thread situation.

**Comment by dillchen** [comment]
@rbennettcw this should not be an issue. TLDR the weight calc will be only for either `MA` or `MB` for any given topic at launch, decided by admin per topic. In the future, it may be a summation across `MA` + `MB`. That is not 

And what you've said seems reasonable. However I want to draw out smth here.

The term "gating" is doing a lot of heavy lifting. 

> The issue I'm seeing is that if a user is reacting to a thread with Topic T, and if Topic T is gated by Group A and Group B, and the user address is accepted in both groups, then there are multiple memberships (MA, MB) and thus multiple stake amounts associated with the address and that thread or comment being reacted to.

Groups are becoming more powerful, and basically what you are describing is "which group is allowed to upvote" in a community or topic, whereas before it was only for "which group is allowed to upvote". They are two distinct permissions

**Designs Below**

These designs treat the score 

https://www.figma.com/file/KOEVNGXStBhmX3NnYq7MxP/Community-Stake?type=design&node-id=1117-208966&mode=design&t=f1wW4jNjbPkK7Ijw-4

<img width="416" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/5334557/0f642fbe-f82f-4aeb-b40a-6436ae206aec">

**Comment by rbennettcw** [comment]
Draft proposal based on discussion:

## Foundational assumptions
  - Admin can assign a group to be the "staking group" `G` for topic `T`
  - Group `G` has a staking token `TK`
  - User (address) can buy stake of token `TK`
    - Stake amount source of truth is on contract

# Event-driven weighted reactions
Being event-driven means using events (messages) as a means of triggering actions. We can use a pubsub class instance which is passed around the app. Pubsub implementation details TBD.

```ts
const pubsub = new PubSub()

pubsub.on('SomeEvent', (eventParams) => {
    // do something
})

pubsub.publish('SomeEvent', { data: 123 })
```

### Data model
- Group should have assigned staking token
    -  Out of scope?
- Topic should have an assigned staking group
    - ✏️ Add `staking_group_id` (number nullable) to `Topics` model
- Membership should reflect total stake for user
    - ✏️ Add `stake_amount` (string nullable) to `Memberships` model
      - Is string since contract uses u256 big number
- Reaction should reflect calculated voting weight
    - ✏️ Add `calculated_voting_weight` (string nullable) to `Reactions` model
      - May also be big number since it's derived from stake amount?
- Pointing strategy should be customizable per group
    - ✏️ Add `pointing_strategy` (string nullable)  to `Groups` model
        - Identifier for the pointing strategy to use for the group
    - ✏️ Add `pointing_params` (JSONB nullable) to `Groups` model
        - User-defined params to be passed to the pointing strategy code implementation
    - ✏️ Create `IPointingStrategy` interface
      - `calculateVotingWeight(state, params) : number`
        - `state` is an object containing current stake amount and other context
        - `params` is the user-defined params
        - Return value is the calculated weight given state and params
    - ✏️ Create `DefaultPointingStrategy` class that implements `IPointingStrategy`
      - Is the default fallback strategy if `group.pointing_strategy` is null
      - Implementation simply maps 1 stake to 1 weight:
        - `calculateVotingWeight(state, params) => state.stake_amount`
    - ✏️ Example custom strategy: `RatioPointingStrategy` class that implements `IPointingStrategy`
      - Chosen if `group.pointing_strategy = 'ratio'`
      - Implementation uses user-defined params for multipler
        - `calculateVotingWeight(state, params) => state.stake_amount / 1e18 * params.multiplier`
        - Simplified example– should actually use BN lib for math ops

### Event handling

The assumption here is that weights are always based on the user's latest stake, not the user's stake at the time the reaction was created.

```ts
pubsub.on('UpdateStake', ({userAddress, groupId, stakeAmount}) => {
  // [A] find membership and update membership.stake_amount
  
  // [B] based on group.pointing_strategy, instantiate the matching class that implements IPointingStrategy
  // [C] invoke IPointingStrategy.calculateVotingWeight(stake amount)
  
  // [D] find all topics that have assigned this group as the staking group
  // [E] find all reactions within those topics
  // [F] bulk update reaction.calculated_voting_weight for all of the user's reactions for the threads/comments within those topics
})
```

From a performance standpoint, the a potential bottleneck would be `[D, E, F]`– finding and updating potentially thousands of reactions. The most performant approach would be to perform a single find-and-update query, perhaps via raw SQL if necessary.

Since the calculated weights are based on the topic's assigned staking group, we may need to do some cleanup when a topic's staking group is changed.

```ts
pubsub.on('UpdateTopicStakingGroup', ({topicId, groupId}) => {
    // recalculate voting weights for reactions within the topic
})
```

### Command
When a reaction is created, the voting weight should be calculated on the fly and persisted to `reaction.calculated_voting_weight`. That requires fetching the current stake amount via TBC, and using `IPointingStrategy`.

### Query
Any routes that return reactions should also return the sum of all reaction weights per user, since that's what will be rendered to the UI.

Specific code changes TBD.
